core/src/main/java/io/questdb/cairo/O3PartitionJob.java

```diff
@@ -1276,69 +1281,145 @@ private static void mergeRowGroup(
+                    if (tableWriterMetadata.isDedupKey(columnIndex) && columnIndex != timestampIndex) {
+                        final int columnSize = !ColumnType.isVarSize(columnType) ? ColumnType.sizeOf(columnType) : -1;
+                        final long columnTop = tableWriter.getColumnTop(partitionTimestamp, columnIndex, rowGroupSize);
+                        long addr = DedupColumnCommitAddresses.setColValues(
+                                dedupColSinkAddr,
+                                dedupColumnIndex++,
+                                columnType,
+                                columnSize,
+                                columnTop
+                        );
+                        if (columnSize > 0) {
+                            DedupColumnCommitAddresses.setColAddressValues(addr, rowGroupBuffers.getChunkDataPtr(bufferIndex));
+                            final long oooColAddress = oooColumns.get(getPrimaryColumnIndex(columnIndex)).addressOf(0);
+                            DedupColumnCommitAddresses.setO3DataAddressValues(addr, oooColAddress);
+                        } else {
+                            DedupColumnCommitAddresses.setColAddressValues(
+                                    addr,
+                                    rowGroupBuffers.getChunkAuxPtr(bufferIndex),
+                                    rowGroupBuffers.getChunkDataPtr(bufferIndex),
+                                    rowGroupBuffers.getChunkDataSize(bufferIndex)
+                            );
+                            MemoryCR oooVarCol = oooColumns.get(getPrimaryColumnIndex(columnIndex));
+                            final long oooVarColAddress = oooVarCol.addressOf(0);
+                            final long oooVarColSize = oooVarCol.addressHi() - oooVarColAddress;
+                            final long oooAuxColAddress = oooColumns.get(getSecondaryColumnIndex(columnIndex)).addressOf(0);
+                            DedupColumnCommitAddresses.setO3DataAddressValues(addr, oooAuxColAddress, oooVarColAddress, oooVarColSize);
+                        }
+                    }
+                }
+
+                dedupRows = Vect.mergeDedupTimestampWithLongIndexIntKeys(
+                        timestampDataPtr,
+                        0,
+                        rowGroupSize - 1,
+                        sortedTimestampsAddr,
+                        mergeRangeLo,
+                        mergeRangeHi,
+                        timestampMergeIndexAddr,
+                        dedupCommitAddresses.getColumnCount(),
+                        DedupColumnCommitAddresses.getAddress(dedupColSinkAddr)
+                );
+            }
+
+            long oldIndexSize = timestampMergeIndexSize;
+            timestampMergeIndexSize = dedupRows * TIMESTAMP_MERGE_ENTRY_BYTES;
+            timestampMergeIndexAddr = Unsafe.realloc(timestampMergeIndexAddr, oldIndexSize, timestampMergeIndexSize, MemoryTag.NATIVE_O3);
```

Comment:
What would happen if this `realloc` call fails? We already changed the `timestampMergeIndexSize` value, so the `Unsafe.free()` call in the subsequent `finally` block will use wrong size and will fail too.

Reply:
addressed